source/dmocks/mocks.d

```diff
@@ -719,8 +719,12 @@ unittest
 {
     Mocker mocker = new Mocker();
     TestClass cl = mocker.mock!(TestClass);
+    
     mocker.replay();
-    assertThrown!ExpectationViolationException(cl.test);
```

Comment:
I prefer the old behavior: fail as soon as the first unexpected call is made.
(But show expected and actual calls in this case.)

`mocker.verify` should only fail in case of missing calls.
(Interestingly enough there seems to be no simple test case for an expected call that is missing.)

Reply:
This breaks `nothrow` functions. If `cl.test` is `nothrow` it would throw and probably cause undefined behaviour. Since there is a chance `nothrow` becomes default (https://forum.dlang.org/thread/sbdrybtyfkxfhxxjgqco@forum.dlang.org), it could hurt in future.

Further consider:

```d
import dmocks.mocks;

class Foo
{
    private TestClass bar;

    this (TestClass bar)
    {
        this.bar = bar;
    }

    void calculateThings()
    {
        try
        {
            bar.test();
        }
        catch (Exception e)
        {
        }
    }
}

class TestClass
{
    string test() { return "test"; }
}


void main()
{
    Mocker mocker = new Mocker();
    TestClass cl = mocker.mock!(TestClass);
    auto foo = new Foo(cl);

    mocker.replay();

    foo.calculateThings();

    mocker.verify;
}
```

`Foo.calculateThings` catches `ExpectationViolationException` (unintended). `mocker.verify` fails but with the same cryptic error message:

> dmocks.util.ExpectationViolationException@../../source/dmocks/util.d(53): 
> Unexpected calls(calls):
> TestClass.test (<unknown>) @system
>
> ----------------
> ../../source/dmocks/repository.d:182 void dmocks.repository.MockRepository.Verify(bool, bool) [0x4dcc4a]
> ../../source/dmocks/mocks.d:55 void dmocks.mocks.Mocker.verify(bool, bool) [0x4da096]
source/app.d:40 _Dmain [0x4bd17c]
> Program exited with code 1

To fix this, the same logic should be implemented in 2 places: mocked objects/functions and `.verify`.
But throwing an unrelated dmocks-Exception from a mocked method messes up the code, that calls the function, since the code doesn't expect this exception.

Reply:
I guess there will be lots of trouble if the exception is postponed to `verify`:
- what if the unexpected call must return some value?
- it's very likely that the function calling the mock fails if it gets `T.init` as return value
- it's even more likely that the remaining unit test fails in checking the result
- both examples would mask the unexpected call
The developer then gets unhelpful failure messages because the unit test is aborted before `verify` is even called.

How about throwing an `Error` in case of an unexpected call? This seems to be a situation that is not recoverable.

Reply:
`Error` breaks chaining and destroys the original Exception if the `Fixture` dtor was run due to an Exception in the unittest.